File: SectionMattr.Samples/Parser.fs
module Parser

open System.Text.Json
open YamlDotNet.Serialization
open YamlDotNet.Serialization.NamingConventions
open SectionMattr

let stringContent =
    @"Content before the sections.

---

More content.

---one
title: One
---

This is section one."

// JSON Parse

type ParsedJsonData = { title: string }

type MyJsonParser() =
    interface IMattrParser<ParsedJsonData> with
        member _.Parse
            (
                section: MattrSection<string>,
                sections: MattrSection<ParsedJsonData> array
            ) : MattrSection<ParsedJsonData> =
            let data: ParsedJsonData = JsonSerializer.Deserialize<ParsedJsonData> section.Data
            NewMattr.AppendData data section

let mattrJSON: Mattr<ParsedJsonData> =
    NewMattr.Sections(stringContent, MyJsonParser())

mattrJSON |> ignore

// YAML Parse

// `YamlDotNet` need this attribute to be able to deserialize
[<CLIMutable>]
type ParsedYamlData = { title: string }

type MyYamlParser() =
    interface IMattrParser<ParsedYamlData> with
        member _.Parse
            (
                section: MattrSection<string>,
                sections: MattrSection<ParsedYamlData> array
            ) : MattrSection<ParsedYamlData> =
            let data: ParsedYamlData =
                (new DeserializerBuilder())
                    .WithNamingConvention(UnderscoredNamingConvention.Instance)
                    .Build()
                    .Deserialize<ParsedYamlData>
                    section.Data

            NewMattr.AppendData data section

let mattrYAML: Mattr<ParsedYamlData> =
    NewMattr.Sections(stringContent, MyYamlParser())

mattrYAML |> ignore

File: SectionMattr.Samples.Csharp/Parser.cs
using System.Text.Json;
using YamlDotNet.Serialization;
using YamlDotNet.Serialization.NamingConventions;

namespace SectionMattr.Samples.Csharp.Parser;

public class ParsedJsonData
{
    public string title { get; set; }
}

class MyJsonParser : IMattrParser<ParsedJsonData>
{
    public MattrSection<ParsedJsonData> Parse(MattrSection<string> section, MattrSection<ParsedJsonData>[] sections)
    {
        ParsedJsonData data = JsonSerializer.Deserialize<ParsedJsonData>(section.Data);
        return NewMattr.AppendData<ParsedJsonData>(data, section);
    }
}

public class ParsedYamlData
{
    public string title { get; set; }
}

class MyYamlParser : IMattrParser<ParsedYamlData>
{
    public MattrSection<ParsedYamlData> Parse(MattrSection<string> section, MattrSection<ParsedYamlData>[] sections)
    {
        ParsedYamlData data = new DeserializerBuilder().WithNamingConvention(UnderscoredNamingConvention.Instance).Build().Deserialize<ParsedYamlData>(section.Data);
        return NewMattr.AppendData<ParsedYamlData>(data, section);
    }
}

public class Parser
{
    static void Main()
    {
        string stringContent =
    @"Content before the sections.

---

More content.

---one
title: One
---

This is section one.";

        // JSON Parse
        Mattr<ParsedJsonData> mattrJSON = NewMattr.Sections(stringContent, new MyJsonParser());

        // YAML Parse
        Mattr<ParsedYamlData> mattrYAML = NewMattr.Sections(stringContent, new MyYamlParser());
    }
}

